packages/modules/gcloud/src/spanner-emulator-container.ts

```diff
@@ -0,0 +1,179 @@
+import { Spanner } from "@google-cloud/spanner";
+import type { IInstance } from "@google-cloud/spanner/build/src/instance";
+import { AbstractStartedContainer, GenericContainer, StartedTestContainer, Wait } from "testcontainers";
+
+const GRPC_PORT = 9010;
+
+/**
+ * SpannerEmulatorContainer runs the Cloud Spanner emulator via the GCloud CLI image.
+ */
+export class SpannerEmulatorContainer extends GenericContainer {
+  private projectId?: string;
+
+  constructor(image: string) {
+    super(image);
+
+    // only gRPC port is supported
+    this.withExposedPorts(GRPC_PORT).withWaitStrategy(Wait.forLogMessage(/.*Cloud Spanner emulator running\..*/, 1));
+  }
+
+  /**
+   * Sets the GCP project ID to use with the emulator.
+   */
+  public withProjectId(projectId: string): this {
+    this.projectId = projectId;
+    return this;
+  }
+
+  public override async start(): Promise<StartedSpannerEmulatorContainer> {
+    const selectedProject = this.projectId ?? "test-project";
+
+    const started = await super.start();
+    return new StartedSpannerEmulatorContainer(started, selectedProject);
+  }
+}
+
+/**
+ * A running Spanner emulator instance with endpoint getters and helper access.
+ */
+export class StartedSpannerEmulatorContainer extends AbstractStartedContainer {
+  constructor(
+    startedTestContainer: StartedTestContainer,
+    private readonly projectId: string
+  ) {
+    super(startedTestContainer);
+  }
+
+  /**
+   * @returns host:port for gRPC.
+   */
+  public getEmulatorGrpcEndpoint(): string {
+    return `${this.getHost()}:${this.getMappedPort(GRPC_PORT)}`;
+  }
+
+  /**
+   * @returns the GCP project ID used by the emulator.
+   */
+  public getProjectId(): string {
+    return this.projectId;
+  }
+
+  /**
+   * @returns a helper for Spanner resource operations against the emulator.
+   */
+  public get helper(): SpannerEmulatorHelper {
+    return new SpannerEmulatorHelper(this);
+  }
+}
+
+/**
+ * Helper class that encapsulates all Spanner client interactions against the emulator.
+ * Clients and configs are lazily instantiated; user must call setAsEmulatorHost().
+ */
+export class SpannerEmulatorHelper {
+  private clientInstance?: Spanner;
+  private instanceAdminClientInstance?: ReturnType<Spanner["getInstanceAdminClient"]>;
+  private databaseAdminClientInstance?: ReturnType<Spanner["getDatabaseAdminClient"]>;
+  private instanceConfigValue?: string;
+
+  constructor(private readonly emulator: StartedSpannerEmulatorContainer) {}
+
+  /**
+   * Must be called by the user to configure the SPANNER_EMULATOR_HOST env var.
+   */
+  public setAsEmulatorHost(): void {
+    process.env.SPANNER_EMULATOR_HOST = this.emulator.getEmulatorGrpcEndpoint();
```

Comment:
Does it make sense to extend the `AbstractGcloudEmulator` class? I see there they're setting the host via a flag: https://github.com/testcontainers/testcontainers-node/blob/4b7b9d0a5ee7f8482936cef8a457039ea616129c/packages/modules/gcloud/src/abstract-gcloud-emulator.ts#L15

If possible, this would be preferable to setting the env var, and then we don't need this setter, or the checks